### ISSUE TITLE
Template type conversion operator not used leading to erroneous compilation error

### ISSUE THREAD

**Issue by llvmbot** [llvm/llvm-project#22146]
|  |  |
| --- | --- |
| Bugzilla Link | [21772](https://llvm.org/bz21772) |
| Version | 3.5 |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@tavianator |

## Extended Description 
The following code looks like valid C++ 11 but fails with error
```
no matching constructor for initialization of 'Ptr<Shape>'
```
It compiles ok on gcc 4.9. Based on example code taken from Stroustrup's The C++ Programming Language 4th edition, pp 764-765.
```cpp
template <typename T>
struct Ptr {
    Ptr(T* pp) : p{pp} { }

    template <typename U>
    explicit operator Ptr<U>(); // Convert Ptr<T> to Ptr<U>
private:
    T* p;
};

template<typename T>
template<typename U>
Ptr<T>::operator Ptr<U>()
{
    return Ptr<U>{p};
}

struct Shape {};
struct Circle : Shape {};

int main()
{
    Circle c;
    Ptr<Circle> pc{&c};
    // The following instantiation causes the
    // error. Ptr<Circle>::operator Ptr<Shape>() should be getting
    // implicitly called on the argument pc, before the result is used
    // to call Ptr<Shape>'s default copy constructor.
    Ptr<Shape> ps{pc};
}
```

**Comment by Endilll** [comment]
Appears to be fixed in Clang 3.7: https://godbolt.org/z/9qW7d4T5d

**Comment by llvmbot** [comment]
@llvm/issue-subscribers-clang-frontend

Author: None (llvmbot)

<details>
|  |  |
| --- | --- |
| Bugzilla Link | [21772](https://llvm.org/bz21772) |
| Version | 3.5 |
| OS | Linux |
| Reporter | LLVM Bugzilla Contributor |
| CC | @DougGregor,@tavianator |

## Extended Description 
The following code looks like valid C++ 11 but fails with error
```
no matching constructor for initialization of 'Ptr&lt;Shape&gt;'
```
It compiles ok on gcc 4.9. Based on example code taken from Stroustrup's The C++ Programming Language 4th edition, pp 764-765.
```cpp
template &lt;typename T&gt;
struct Ptr {
    Ptr(T* pp) : p{pp} { }

    template &lt;typename U&gt;
    explicit operator Ptr&lt;U&gt;(); // Convert Ptr&lt;T&gt; to Ptr&lt;U&gt;
private:
    T* p;
};

template&lt;typename T&gt;
template&lt;typename U&gt;
Ptr&lt;T&gt;::operator Ptr&lt;U&gt;()
{
    return Ptr&lt;U&gt;{p};
}

struct Shape {};
struct Circle : Shape {};

int main()
{
    Circle c;
    Ptr&lt;Circle&gt; pc{&amp;c};
    // The following instantiation causes the
    // error. Ptr&lt;Circle&gt;::operator Ptr&lt;Shape&gt;() should be getting
    // implicitly called on the argument pc, before the result is used
    // to call Ptr&lt;Shape&gt;'s default copy constructor.
    Ptr&lt;Shape&gt; ps{pc};
}
```
</details>
